src/jit/flowgraph.cpp

```diff
@@ -22445,6 +22445,28 @@ GenTreePtr Compiler::fgInlinePrependStatements(InlineInfo* inlineInfo)
 
 void Compiler::fgInlineAppendStatements(InlineInfo* inlineInfo, BasicBlock* block, GenTreePtr stmtAfter)
 {
+    // If this inlinee was passed a generic context but didn't look at
+    // it, we can decrement the "generic context was used" ref count.
+    if ((inlineInfo->inlineCandidateInfo->methInfo.args.callConv & CORINFO_CALLCONV_PARAMTYPE) != 0)
+    {
+        // Did the context require the caller to perform a runtime lookup?
+        if (inlineInfo->inlineCandidateInfo->exactContextNeedsRuntimeLookup)
+        {
+            // Fetch the temp for the generic context
+            const unsigned typeCtxtArg = inlineInfo->typeContextArg;
+            const unsigned tmpNum      = inlineInfo->lclTmpNum[typeCtxtArg];
+
+            // Was it used in the inline body?
+            if (tmpNum == BAD_VAR_NUM)
+            {
+                // No, so the runtime lookup is not needed.
+                JITDUMP("Inlinee ignores runtime lookup generics context\n");
+                assert(lvaGenericsContextUseCount > 0);
+                lvaGenericsContextUseCount--;
```

Comment:
Yeah, I figured this would be confusing.

The current code assumes that the generic context must be kept alive whenever `getRuntimeContextTree` is called, by setting `lvaGenericsContextIsUsed` true. This flag is checked over in `lvaReportParamTypeArg` and `lvaKeepAliveAndReportThis`. Note because we won't inline methods that internally request runtime lookups, in any given method there is at most one generics context in play and it is either the this pointer or the type parameter, and all these runtime lookup trees start from that context.

We don't cache/memoize these lookups so each importer operation that requires on generates a new tree. Given the potential indirect/helper lookups required each "tree" is really 3 statements.

Some subset of these tree-inspiring importations come from calls, where the result of the lookup is passed as an explicit type parameter. For inlinable calls, this happens when importing calls with CORINFO_CALLCONV_PARAMTYPE via `impTokenToHandle`. And some subset of those calls may end up getting inlined. And some subset of those inlined calls will ignore the generic context. And when that happens the results of the tree lookup are unused.

We also don't remove any dead / unreachable code throughout importation and inlining. So at the end of inlining, we have created N lookup trees, N-k of which we know for sure are ignored.

Subsequently the optimizer runs and will delete those k trees as dead code. It may also CSE trees and/or delete even more trees as dead code. So after optimization there are N-k-z trees.

Currently we will always report the context as live if N > 0. This change introduces ref counting to try and spot the case where N == k and if so, avoid keeping the generics context -- the root of all the runtime lookup trees -- live and reportable in the gc info.

A better fix would be to somehow realize what the existence of these trees implies and get the cases where N == k + z too. However I don't know of any examples where z > 0 -- maybe you'd see this from an unused ldvirtfn.

There's a deeper issue here that deserves more consideration -- the question of why the context needs to be kept live in the first place, above and beyond the explicit VM requests (eg for generic catch type filters). 

I don't claim to fully understand this case (the code refers to "collectible classes"). If I had to guess I'd say it was to guard against cases where we have a last use of the generic parameter value at a call that ignores the parameter, so that when the callee is active its class can't be GCd. If so then that contract is still preserved here.

@jkotas is that last bit more or less right?